File: src/Integra/Evaluator.hs
module Integra.Evaluator (eval, evalWithX, showComplex, niceShow) where

import Integra.AST (Expr(..))
import Integra.Special (gamma, erf)
import Data.Complex (Complex(..), realPart, imagPart, magnitude)

eval :: Complex Double -> Complex Double -> Expr -> Complex Double
eval x ansVal = go
  where
    go (Num n)       = n :+ 0
    go Pi            = pi :+ 0
    go E             = exp 1 :+ 0
    go Tau           = (2 * pi) :+ 0
    go Phi           = ((1 + sqrt 5) / 2) :+ 0
    go I             = 0 :+ 1
    go Var           = x
    go Ans           = ansVal
    go (Add l r)     = go l + go r
    go (Sub l r)     = go l - go r
    go (Mul l r)     = go l * go r
    go (Div l r)     = go l / go r
    go (Pow l r)     = go l ** go r
    go (SinE   e)    = sin (go e)
    go (CosE   e)    = cos (go e)
    go (TanE   e)    = tan (go e)
    go (CscE   e)    = 1 / sin (go e)
    go (SecE   e)    = 1 / cos (go e)
    go (CotE   e)    = 1 / tan (go e)
    go (AsinE  e)    = asin (go e)
    go (AcosE  e)    = acos (go e)
    go (AtanE  e)    = atan (go e)
    go (AcscE  e)    = asin (1 / go e)
    go (AsecE  e)    = acos (1 / go e)
    go (AcotE  e)    = atan (1 / go e)
    go (SinhE  e)    = sinh (go e)
    go (CoshE  e)    = cosh (go e)
    go (TanhE  e)    = tanh (go e)
    go (CschE  e)    = 1 / sinh (go e)
    go (SechE  e)    = 1 / cosh (go e)
    go (CothE  e)    = 1 / tanh (go e)
    go (AsinhE e)    = asinh (go e)
    go (AcoshE e)    = acosh (go e)
    go (AtanhE e)    = atanh (go e)
    go (LogE   e)    = log (go e)
    go (Log2E  e)    = log (go e) / log (2 :+ 0)
    go (Log10E e)    = log (go e) / log (10 :+ 0)
    go (ExpE   e)    = exp (go e)
    go (SqrtE  e)    = sqrt (go e)
    go (AbsE   e)    = magnitude (go e) :+ 0
    go (SignE  e)    = signum (go e)
    go (FloorE e)    = (fromIntegral (floor (asReal (go e)) :: Int) :+ 0)
    go (CeilE  e)    = (fromIntegral (ceiling (asReal (go e)) :: Int) :+ 0)
    go (RoundE e)    = (fromIntegral (round (asReal (go e)) :: Int) :+ 0)
    go (GammaE e)    = gamma (asReal (go e)) :+ 0
    go (ErfE   e)    = erf (asReal (go e)) :+ 0
    go (ConjE  e)    = let z = go e in realPart z :+ (-imagPart z)
    go (ReE    e)    = realPart (go e) :+ 0
    go (ImE    e)    = imagPart (go e) :+ 0

    asReal z
        | abs (imagPart z) < 1e-12 = realPart z
        | otherwise = error "Expected a real number, got complex"

evalWithX :: Double -> Expr -> Complex Double
evalWithX x = eval (x :+ 0) 0

------------------------------------------------------------
-- DISPLAY HELPERS
------------------------------------------------------------

niceShow :: Double -> String
niceShow x
    | isNaN x           = "undefined"
    | isInfinite x      = if x > 0 then "∞" else "-∞"
    | x == 0            = "0"
    | abs x >= 1e12
      || (abs x < 1e-8 && abs x > 0) = show x
    | otherwise =
        let s  = show x
            s' = if '.' `elem` s
                 then reverse (dropWhile (== '0') (reverse s))
                 else s
        in if last s' == '.' then init s' else s'

showComplex :: Complex Double -> String
showComplex z
    | isNaN (realPart z) || isInfinite (realPart z)
        = niceShow (realPart z)
    | isNaN (imagPart z) || isInfinite (imagPart z)
        = niceShow (imagPart z) ++ "i"
    | abs (imagPart z) < 1e-12
        = niceShow (realPart z)
    | abs (realPart z) < 1e-12
        = showImag (imagPart z)
    | imagPart z > 0
        = niceShow (realPart z) ++ " + " ++ showImag (imagPart z)
    | otherwise
        = niceShow (realPart z) ++ " - " ++ showImag (-imagPart z)
  where
    showImag x
        | abs (abs x - 1) < 1e-12 = if x > 0 then "i" else "-i"
        | otherwise               = niceShow x ++ "i"

File: wasm/src/evaluator.rs
use crate::ast::AST;
use crate::complex::C;

pub fn eval(ast: &AST, x: C, ans: C) -> C {
    match ast {
        AST::Num(n) => C::from(*n),
        AST::Pi => C::from(std::f64::consts::PI),
        AST::E => C::from(std::f64::consts::E),
        AST::Tau => C::from(std::f64::consts::TAU),
        AST::Phi => C::from((1.0 + 5.0_f64.sqrt()) / 2.0),
        AST::I => C::new(0.0, 1.0),
        AST::Var => x,
        AST::Ans => ans,
        AST::Add(a, b) => eval(a, x, ans) + eval(b, x, ans),
        AST::Sub(a, b) => eval(a, x, ans) - eval(b, x, ans),
        AST::Mul(a, b) => eval(a, x, ans) * eval(b, x, ans),
        AST::Div(a, b) => eval(a, x, ans) / eval(b, x, ans),
        AST::Pow(a, b) => eval(a, x, ans).pow(eval(b, x, ans)),
        AST::Gt(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if va.re > vb.re { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Ge(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if va.re >= vb.re { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Lt(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if va.re < vb.re { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Le(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if va.re <= vb.re { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Eqq(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if (va.re - vb.re).abs() < 1e-12 { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Neq(a, b) => {
            let va = eval(a, x, ans);
            let vb = eval(b, x, ans);
            if (va.re - vb.re).abs() >= 1e-12 { C::from(1.0) } else { C::from(0.0) }
        }
        AST::Sin(a) => eval(a, x, ans).sin(),
        AST::Cos(a) => eval(a, x, ans).cos(),
        AST::Tan(a) => eval(a, x, ans).tan(),
        AST::Csc(a) => eval(a, x, ans).csc(),
        AST::Sec(a) => eval(a, x, ans).sec(),
        AST::Cot(a) => eval(a, x, ans).cot(),
        AST::Asin(a) => eval(a, x, ans).asin(),
        AST::Acos(a) => eval(a, x, ans).acos(),
        AST::Atan(a) => eval(a, x, ans).atan(),
        AST::Acsc(a) => eval(a, x, ans).acsc(),
        AST::Asec(a) => eval(a, x, ans).asec(),
        AST::Acot(a) => eval(a, x, ans).acot(),
        AST::Sinh(a) => eval(a, x, ans).sinh(),
        AST::Cosh(a) => eval(a, x, ans).cosh(),
        AST::Tanh(a) => eval(a, x, ans).tanh(),
        AST::Csch(a) => eval(a, x, ans).csch(),
        AST::Sech(a) => eval(a, x, ans).sech(),
        AST::Coth(a) => eval(a, x, ans).coth(),
        AST::Asinh(a) => eval(a, x, ans).asinh(),
        AST::Acosh(a) => eval(a, x, ans).acosh(),
        AST::Atanh(a) => eval(a, x, ans).atanh(),
        AST::Log(a) => eval(a, x, ans).ln(),
        AST::Log2(a) => eval(a, x, ans).log2(),
        AST::Log10(a) => eval(a, x, ans).log10(),
        AST::Log1p(a) => eval(a, x, ans).log1p(),
        AST::Exp(a) => eval(a, x, ans).exp(),
        AST::Expm1(a) => eval(a, x, ans).expm1(),
        AST::Sqrt(a) => eval(a, x, ans).sqrt(),
        AST::Cbrt(a) => eval(a, x, ans).cbrt(),
        AST::Abs(a) => C::from(eval(a, x, ans).abs()),
        AST::Sign(a) => {
            let v = eval(a, x, ans);
            if v.re > 0.0 { C::from(1.0) } else if v.re < 0.0 { C::from(-1.0) } else { C::from(0.0) }
        }
        AST::Floor(a) => C::from(eval(a, x, ans).re.floor()),
        AST::Ceil(a) => C::from(eval(a, x, ans).re.ceil()),
        AST::Round(a) => C::from(eval(a, x, ans).re.round()),
        AST::Gamma(a) => eval(a, x, ans).gamma(),
        AST::Erf(a) => eval(a, x, ans).erf(),
        AST::Conj(a) => eval(a, x, ans).conj(),
        AST::Re(a) => C::from(eval(a, x, ans).re),
        AST::Im(a) => C::from(eval(a, x, ans).im),
    }
}

pub fn eval_real(ast: &AST, x: f64) -> f64 {
    eval(ast, C::from(x), C::from(0.0)).re
}

pub fn eval_c(ast: &AST, x: f64) -> C {
    eval(ast, C::from(x), C::from(0.0))
}
